pkg/controller/volume/persistentvolume/scheduler_binder.go

```diff
@@ -0,0 +1,414 @@
+			return true, err
+		}
+		if dirty {
+			err = b.pvCache.Assume(newPV)
+			if err != nil {
+				b.revertAssumedPVs(newBindings)
+				return true, err
+			}
+
+			newBindings = append(newBindings, &bindingInfo{pv: newPV, pvc: binding.pvc})
+		}
+	}
+
+	if len(newBindings) == 0 {
+		// Update cached bindings even if empty. This can happen if we previously updated
+		// the PV object and are waiting for the PV controller to finish binding
+		//
+		// TODO: can this be optimized so that we don't have to spawn a go routine for binding
+		// that will exit immediately?
+		glog.V(4).Infof("AssumePodVolumes: PVs already assumed")
+	}
+	b.podBindingCache.UpdateBindings(podName, nodeName, newBindings)
+
+	return true, nil
+}
+
+// BindPodVolumes gets the cached bindings in podBindingCache and makes the API update for those PVs.
+func (b *volumeBinder) BindPodVolumes(assumedPod *v1.Pod) error {
+	podName := getPodName(assumedPod)
+
+	glog.V(4).Infof("BindPodVolumes for pod %q", podName)
+
+	bindings := b.podBindingCache.GetBindings(podName, assumedPod.Spec.NodeName)
+
+	// Do the actual prebinding. Let the PV controller take care of the rest
+	// There is no API rollback if the actual binding fails
+	for i, bindingInfo := range bindings {
+		_, err := b.ctrl.updateBindVolumeToClaim(bindingInfo.pv, bindingInfo.pvc, false)
+		if err != nil {
+			// only revert assumed cached updates for volumes we haven't successfully bound
+			b.revertAssumedPVs(bindings[i:])
+			return err
+		}
+	}
+
+	return nil
+}
+
+func getPodName(pod *v1.Pod) string {
+	return pod.Namespace + "/" + pod.Name
+}
+
+func getPVCName(pvc *v1.PersistentVolumeClaim) string {
+	return pvc.Namespace + "/" + pvc.Name
+}
+
+func (b *volumeBinder) isVolumeBound(namespace string, vol *v1.Volume, checkFullyBound bool) (bool, *v1.PersistentVolumeClaim, error) {
+	if vol.PersistentVolumeClaim == nil {
+		return true, nil, nil
+	}
+
+	pvcName := vol.PersistentVolumeClaim.ClaimName
+	pvc, err := b.pvcCache.PersistentVolumeClaims(namespace).Get(pvcName)
+	if err != nil || pvc == nil {
+		return false, nil, fmt.Errorf("error getting PVC %q: %v", pvcName, err)
+	}
+
+	pvName := pvc.Spec.VolumeName
+	if pvName != "" {
+		if checkFullyBound {
+			if metav1.HasAnnotation(pvc.ObjectMeta, annBindCompleted) {
+				glog.V(5).Infof("PVC %q is fully bound to PV %q", getPVCName(pvc), pvName)
+				return true, pvc, nil
+			} else {
+				glog.V(5).Infof("PVC %q is not fully bound to PV %q", getPVCName(pvc), pvName)
+				return false, pvc, nil
+			}
+		}
+		glog.V(5).Infof("PVC %q is bound or prebound to PV %q", getPVCName(pvc), pvName)
+		return true, pvc, nil
+	}
+
+	glog.V(5).Infof("PVC %q is not bound", namespace, getPVCName(pvc))
+	return false, pvc, nil
+}
+
+// arePodVolumesBound returns true if all volumes are fully bound
+func (b *volumeBinder) arePodVolumesBound(pod *v1.Pod) bool {
+	for _, vol := range pod.Spec.Volumes {
+		if isBound, _, _ := b.isVolumeBound(pod.Namespace, &vol, true); !isBound {
```

Comment:
Why `true` here? It's called by `AssumePodVolumes` and if there are some PVCs waiting for `annBindCompleted`, then `AssumePodVolumes` will try to update `pvc.spec.volumeName`. But it's already updated...

IMO, `false` would do the same job, saving some cycles in `AssumePodVolumes` and one `BindPodVolumes` call. There is nothing to do, scheduler is just waiting for PV controller to finish the binding. Am I missing something?

Reply:
Right now, `bindVolumesWorker` in the scheduler is the only one that adds the pod back to the scheduler queue, so everything must pass through that currently, even if we're just waiting and there's nothing to do.  I could short circuit it in the `assumeVolumes` as well to save some cycles.  I had left that as a future TODO but I can try to do it now.